git/pre-commit

```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
+set -eu
+
+# This file will exist in $project_root/.git/hooks/, so we need to move back up
+# to $project_root. This apparent mess needs to happen because macOS doesn't
+# have `readlink -f`.
+declare src="$0"
+declare dir=''
+
+while [ -h "$src" ]; do
+  dir="$(cd -P "$( dirname "$src")" && pwd)"
+  src="$(readlink "$src")"
+  [[ "$src" != /* ]] && src="$dir/$src"
+done
+
+cd -P "$(dirname "$src")/../../"
+
+##### This is the actual pre-commit hook. #####
+
+# If the developer has a virtual environment in the canonical location, then
+# this will attempt to use that. Otherwise, assume the necessary things are
+# installed globally.
+declare -r venv='./venv/bin/activate'
+if [ -f "$venv" ]; then
+  source "$venv"
+fi
+
+make lint
```

Comment:
Running the docs-lint task every time adds considerable delay here. Compare:

```
$ time make lint > /dev/null

real    0m11.934s
user    0m14.372s
sys     0m0.200s
$ time sh -c 'make flake8 ; make html-lint' > /dev/null

real    0m1.923s
user    0m4.516s
sys     0m0.096s
```

The extra ten seconds will add a ton of time when rebasing. Maybe skip the docs-lint step for interactive work and leave it in CI?

Reply:
This doesn't run when rebasing, so I think that's not an issue.

Reply:
In theory, we could actually do a parse of `git status --porcelain` and then only run the relevant tests, though that would quickly become a pain to maintain.